src/utils/AutoControlledComponent-slim.js

```diff
@@ -0,0 +1,45 @@
+/* eslint-disable */
+import _ from 'lodash'
+import { Component } from 'react'
+
+const toDefault = (prop) => `default${prop[0].toUpperCase() + prop.slice(1)}`
+
+export default class AutoControlledComponent extends Component {
+  componentWillMount() {
+    if (super.componentWillMount) super.componentWillMount()
+    const { autoControlledProps } = this.constructor
+
+    this.state = _.transform(autoControlledProps, (res, prop) => {
+      const defaultProp = toDefault(prop)
+
+      // apply default props if they exist
+      res[prop] = this.props[defaultProp in this.props ? defaultProp : prop]
+    }, {})
+  }
+
+  componentWillReceiveProps(nextProps) {
+    super.componentWillReceiveProps(nextProps)
+
+    // props always win, update state with all auto controlled prop
+    const newState = _.pick(nextProps, this.constructor.autoControlledProps)
+    if (!_.isEmpty(newState)) this.setState(newState)
+  }
+
+  /**
+   * Safely attempt to set state for props that might be controlled by the user.
+   * Second argument is a state object that is always passed to setState.
+   * @param {object} maybeState State that corresponds to controlled props.
+   * @param {object} [state] Actual state, useful when you also need to setState.
```

Comment:
TL;DR - allows single render for setting state.

You can see it in action all over in the new dropdown, like [here](https://github.com/TechnologyAdvice/stardust/pull/226/files#diff-609214949026e594dbfd98481664c928R443).

Also, try set state is really just smart set state so it makes sense to set actual state along with maybe state in a single method.  I contemplated parsing a single state object into maybe state and normal state but this was cleaner.